tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -14308,6 +14313,91 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC, Expr *BitWidth,
+          llvm::APSInt Arg1val = SpecDecl->getTemplateArgs()[0].getAsIntegral();
+          bool IsForceOMM2State =
+              Arg1val.getZExtValue() & (unsigned)DXIL::RayFlag::ForceOMM2State;
+          if (numArgs == 1 && IsForceOMM2State) {
+            Diag(D.getLocStart(), diag::warn_hlsl_unexpected_rayquery_flag);
+            result = false;
+          }
+
+          // here we're guaranteed 2 args
+          else {
+            // ensure that if the second template argument has a non-zero value,
+            // the shader model is at least 6.9
+            const TemplateArgument *Args = TST->getArgs();
+            auto Arg2 = Args[1];
+            llvm::APSInt Arg2val =
+                SpecDecl->getTemplateArgs()[1].getAsIntegral();
+            std::string profile = getLangOpts().HLSLProfile;
+            const ShaderModel *SM =
+                hlsl::ShaderModel::GetByName(profile.c_str());
+
+            if (Arg2val.getZExtValue() != 0 && !SM->IsSMAtLeast(6, 9)) {
+              // if it's an enum value, then emit
+              // warn_hlsl_enum_type_not_allowed otherwise emit
+              // warn_hlsl_unexpected_value_for_ray_query_flags_template_arg
+
+              // we can tell that it's an enum value if it has this AST
+              // structure:
+              //`- ImplicitCastExpr 0x2d9b604a870 'unsigned int' <LValueToRValue
+              //> |   `- DeclRefExpr 0x2d9b604a760 'const unsigned int' lvalue
+              // Var 0x2d9b6005840 'RAY_FLAG_FORCE_OMM_2_STATE' 'const unsigned
+              // int'
+
+              if (auto *castExpr = dyn_cast<ImplicitCastExpr>(
+                      Arg2.getAsExpr()->IgnoreParens())) {
+                // Now check if the sub-expression is a DeclRefExpr
+                Expr *subExpr = castExpr->getSubExpr();
+                if (auto *DRE = dyn_cast<DeclRefExpr>(subExpr)) {
+                  Diag(DRE->getDecl()->getLocStart(),
+                       diag::warn_hlsl_enum_type_not_allowed)
+                      << DRE->getDecl()->getIdentifier()->getName()
+                      << SM->GetName() << "6.9";
+                }
+                // otherwise, it could be an integer literal
+                else if (auto *IL = dyn_cast<IntegerLiteral>(subExpr)) {
+                  Diag(
+                      D.getLocStart(),
+                      diag::
+                          warn_hlsl_unexpected_value_for_ray_query_flags_template_arg);
+                }
+              }
+              result = false;
+            }
+            if (IsForceOMM2State) {
+              // additionally check that the expected flag is given in the
+              // second arg
+              if (!(Arg2val.getZExtValue() &
+                    (unsigned)DXIL::RayQueryFlag::AllowOpacityMicromaps)) {
+                Diag(D.getLocStart(), diag::warn_hlsl_unexpected_rayquery_flag);
+                result = false;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
```

Comment:
Nesting this deep makes me wonder if there's some extraction or early returns that could be done here to try and make this code easier to follow.  I notice that this is the deepest nesting I could find on a quick survey through this file.

Reply:
I've done some early returns to help make this more flat.